### PR TITLE
feat(depends,makedepends,optdepends): add version constraints

### PR DESCRIPTION
**Pull request by Elsie19**
## Purpose

We can now do things like:
```bash
depends=("foo>=1.2.3 | bar<=1.2.3")
makedepends=("libfoo=1.0.1")
optdepends=("libidk:i386<5.2.3 | libidk:i386>1.2.5: provides libidk support")
```

## Approach

Now there is a new file `version-constraints.sh` that houses all the functions for this:

* `dep_const.apt_compare_to_constraints` will compare a versioned string against either the locally installed pkg if available and if not, the remote one and return 0 if the condition is satisfied and 1 if not.
* `dep_const.pipe_split` will split a pipe delimited string into an array.
* `dep_const.comma_array` takes a bash array and prints it in a debian control file format.
* `dep_const.split_name_and_version` splits a versioned string into an array with the name and version.
* `dep_const.get_pipe` returns the best package to use based off of an alt dependency string.
* `dep_const.strip_description` strips out the description of a package.
* `dep_const.format_version` converts a pacscript version constrained string into a control file compatible one.
* `dep_const.format_control` converts a pacscript version array into a control file compatible list (excluding commas).

## TODO

- [x] Wait for #1017.
- [x] Edit `checks.sh` and add some stuff.
- [x] Have a check to ensure the constraints can be met before building.

## Other

1. We cannot do this for `pacdeps` however because that would require both knowing the `gives` and `pkgver` beforehand which would be a security risk as we would need to remotely source the pacdep script. We *could* query the website for stuff, but I don't want pacstall to depend on the website.
3. We can still continue using unspecified package versions, this is optional.

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

File: misc/scripts/version-constraints.sh [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+#
+# @example
+#	foo=("opt:amd64>=1.2.3 | bruh:arm64<1.2.0: optdepends string" "blorg>=1.2.3 | larp<=0.0.1")
+#   dep_const.format_control foo out
+#   declare -p out
+#   declare -a out=([0]="opt:amd64 (>= 1.2.3) | bruh:arm64 (<< 1.2.0)" [1]="blorg (>= 1.2.3) | larp (<= 0.0.1)")
+#
+# @arg $1 string An array name.
+# @arg $2 string An array name to output to.
+function dep_const.format_control() {
```
[the remark this answers is not in the document]
Comment (Elsie19) [reply misc/scripts/version-constraints.sh:193]: Stupid review, who let an LLM here?

**Comment by Elsie19** [comment]
I forgot I still need to integrate this into pacstall lol. I will after #1022, #1016, and maybe #1017.

**Comment by oklopfer** [comment]
https://github.com/pacstall/pacstall-programs/blob/master/packages/android-studio/android-studio.pacscript#L6
how will this be affected

**Comment by Elsie19** [comment]
How would it?

**Comment by Elsie19** [comment]
How would it?

**Comment by oklopfer** [comment]
> How would it?

`"libbz2-1.0:i386"`
would this still work alright

**Comment by Xdavius** [comment]
@Elsie19 

Hi.

Any news for this feature ?

Thank you

**Comment by Elsie19** [comment]
Still working on it

**Comment by Elsie19** [comment]
There is still a bug in `dep_const.get_pipe()` where if both sides of the pipe cannot be satisfied, nothing will be returned. That case needs to be handled.

**Review by oklopfer — APPROVED** [review]

**Review by oklopfer — DISMISSED** [review]

File: misc/scripts/checks.sh [hunk 2]
```diff
@@ -210,45 +210,37 @@ function lint_maintainer() {
...
-                fancy_message error "'depends' index '${idx}' cannot be empty"
-                ret=1
-            fi
-            ((idx++))
-        done
-    fi
-    return "${ret}"
+function lint_pipe_check() {
+    perl -ne 'exit 1 unless /^(?:[^\s|:]+(?::[^\s|:]+)?\s\|\s)+[^\s|:]+(?::[^\s|:]+)?(?::\s[^|:]+)?(?<!\s)$/' <<< "$1"
+    return $?
```
Comment (Elsie19) [inline misc/scripts/checks.sh:215]:
```suggestion
    perl -ne 'exit 1 unless /^(?:[^\s|:]+(?::[^\s|:]+)?\s\|\s)+[^\s|:]+(?::[^\s|:]+)?(?::\s[^|:]+)?(?<!\s)$/' <<< "$1"
```

**Review by oklopfer — APPROVED** [review]

**Pull request merged**